plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -596,27 +601,158 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        String bucketName = getBucketName(uri);
+        String key = keyFromPath(path);
+        ObjectMetadata s3ObjectMetadata = getS3ObjectMetadata(path, bucketName, key);
+        ObjectMetadata s3DirectoryMetadata = null;
+        if (!key.isEmpty()) {
+            s3DirectoryMetadata = getS3ObjectMetadata(path, bucketName, key + PATH_SEPARATOR);
+        }
+
+        if (recursive) {
+            boolean deleted = false;
+            if (s3ObjectMetadata == null ||
+                    s3DirectoryMetadata == null ||
+                    MediaType.parse(s3DirectoryMetadata.getContentType()).is(DIRECTORY_MEDIA_TYPE)) {
+                deleted = deletePrefix(path);
+            }
+            if (key.endsWith(DIRECTORY_SUFFIX)) {
+                Path strippedPath = new Path(stripDirectorySuffix(path.toString()));
+                deleted = deletePrefix(strippedPath) || deleted;
             }
+            if (s3ObjectMetadata != null) {
+                deleted = deleteObject(key) || deleted;
+            }
+            if (s3DirectoryMetadata != null) {
+                if (MediaType.parse(s3DirectoryMetadata.getContentType()).is(DIRECTORY_MEDIA_TYPE) ||
+                        key.endsWith(DIRECTORY_SUFFIX)) {
+                    String directoryKey = (key.endsWith(DIRECTORY_SUFFIX) ? stripDirectorySuffix(key) : key) + PATH_SEPARATOR;
```

Comment:
When keys end with `_$folder$` exist as directories, don't we want to delete that object specifically and not the "stripped" object equivalent? It seems like above on line 609, we would assign `s3DirectoryMetadata` to `s3://bucket/directory_$folder$/` (not stripped, but with trailing slash added).

Reply:
> When keys end with _$folder$ exist as directories, don't we want to delete that object specifically and not the "stripped" object equivalent.

I tried to do this initially :) but S3 seems to be effective on doing the deletion only when deleting the stripped object key.

Reply:
I have now two test cases covering this specific question:

- `testDeleteRecursivelyPrefixWithDirectorySuffixContainingMultipleObjects`
- `testDeleteRecursivelyPrefixCorrespondingToDirectoryObjectWithDirectorySuffixContainingMultipleObjects`

Reply:
> I tried to do this initially :) but S3 seems to be effective on doing the deletion only when deleting the stripped object key.

I took a look at the tests in question, seems like this is an artifact of the way the test is written and not really the S3 behavior we're trying to test.